Lib/_pyrepl/utils.py

```diff
@@ -1,13 +1,50 @@
+from __future__ import annotations
+import builtins
+import functools
+import keyword
 import re
+import token as T
+import tokenize
 import unicodedata
-import functools
+import _colorize
+
+from collections import deque
+from io import StringIO
+from tokenize import TokenInfo as TI
+from typing import Iterable, Iterator, Match, NamedTuple, Self
 
 from .types import CharBuffer, CharWidths
 from .trace import trace
 
 ANSI_ESCAPE_SEQUENCE = re.compile(r"\x1b\[[ -@]*[A-~]")
 ZERO_WIDTH_BRACKET = re.compile(r"\x01.*?\x02")
 ZERO_WIDTH_TRANS = str.maketrans({"\x01": "", "\x02": ""})
+IDENTIFIERS_AFTER = {"def", "class"}
+BUILTINS = {str(name) for name in dir(builtins) if not name.startswith('_')}
+
+
+class Span(NamedTuple):
+    """Span indexing that's inclusive on both ends."""
+
+    start: int
+    end: int
+
+    @classmethod
+    def from_re(cls, m: Match[str], group: int | str) -> Self:
+        re_span = m.span(group)
+        return cls(re_span[0], re_span[1] - 1)
+
+    @classmethod
+    def from_token(cls, token: TI, line_len: list[int]) -> Self:
+        return cls(
+            line_len[token.start[0] - 1] + token.start[1],
```

Comment:
This assumes token.start[0] - 1 is a valid index. But IIRC if the tokenizer emits a token on an empty or incomplete line (e.g., EOF inside a triple-quoted string), this can index beyond line_len, no?

Reply:
`token.start[0]` is the line no. The tokenizer uses 1-indexing for real lines.

And before those it only ever emits ENCODING when you're using `tokenize.tokenize()` (i.e. bytes-based tokenization). This is never emitted by `tokenize.generate_tokens()` (i.e. Unicode-based tokenization), but even if it were, it is a case where start and end are both `(0, 0)` so we have a check for that  right at the top of the for loop in `gen_colors_from_token_stream`.

Reply:
ENDMARKER similarly emits start == end.